### PR TITLE
Switch to Provider API

### PR DESCRIPTION
**Pull request by Larusso**
## Description

The current plugin source used the old and internal convention mapping logic. It seems it is still supported up to gradle `6.4`. But understanding and using this API is not straight forward. This patch moves to the `Provider` API introduced in gradle `4.4`. We will lose support for some earlier gradle `4.x` versions but the code becomes easier to read and maintain. I adjusted the `gradle_check_versions.sh` file to test all gradle releases up to `6.4`. All worked well except some internal defaults when it comes to work with the `Provider` API and some breaking changes with gradle `6.x`. I fixed those with a `compat` helper class and resolve the API break at runtime. Since the goal is to move to gradle `6.x` by the end of the year I see no major issue with this.

## Changes

* ![IMPROVE] switch to `Provider` API
* ![UPDATE] ![GRADLE] base version to `4.10.2`

[NEW]:https://atlas-resources.wooga.com/icons/icon_new.svg "New"
[ADD]:http://resources.atlas.wooga.com/icons/icon_add.svg "Add"
[IMPROVE]:http://resources.atlas.wooga.com/icons/icon_improve.svg "IMPROVE"
[CHANGE]:http://resources.atlas.wooga.com/icons/icon_change.svg "Change"
[FIX]:http://resources.atlas.wooga.com/icons/icon_fix.svg "Fix"
[UPDATE]:http://resources.atlas.wooga.com/icons/icon_update.svg "Update"

[BREAK]:http://resources.atlas.wooga.com/icons/icon_break.svg "Remove"
[REMOVE]:http://resources.atlas.wooga.com/icons/icon_remove.svg "Remove"
[IOS]:http://resources.atlas.wooga.com/icons/icon_iOS.svg "iOS"
[ANDROID]:http://resources.atlas.wooga.com/icons/icon_android.svg "Android"
[GRADLE]:https://atlas-resources.wooga.com/icons/icon_gradle.svg "GRADLE"
[WEBGL]:http://resources.atlas.wooga.com/icons/icon_webGL.svg "Web:GL"

### CODE REVIEW

File: src/main/groovy/wooga/gradle/appcenter/tasks/AppCenterUploadTask.groovy [hunk 1]
```diff
@@ -182,42 +136,46 @@ class AppCenterUploadTask extends ConventionTask {
...
+
+    AppCenterUploadTask() {
+        def projectLayout = new ProjectLayout(project)
+        apiToken = project.objects.property(String)
+        owner = project.objects.property(String)
+        buildVersion = project.objects.property(String)
+        releaseId = project.objects.property(Integer)
+        applicationIdentifier = project.objects.property(String)
+        releaseNotes = project.objects.property(String)
+        destinations = project.objects.listProperty(Map)
```
Comment (alexandershestakov) [inline src/main/groovy/wooga/gradle/appcenter/tasks/AppCenterUploadTask.groovy:172]: This looks funky, just curious, why do you need to do that?

File: src/main/groovy/wooga/gradle/compat/ProjectLayout.groovy [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    }
+
+    RegularFileProperty fileProperty() {
+        (RegularFileProperty) compatFactory.invokeMethod("fileProperty", null)
+    }
+
+    RegularFileProperty fileProperty(Provider<? extends RegularFile> var1) {
+        (RegularFileProperty) compatFactory.invokeMethod("fileProperty", var1)
+    }
+}
```
Comment (alexandershestakov) [inline src/main/groovy/wooga/gradle/compat/ProjectLayout.groovy:37]: Why is this needed?

**Review by alexandershestakov — APPROVED** [review]
Looks cleaner, lgtm 👍 

Just asking a couple of questions out of curiousity.

File: src/main/groovy/wooga/gradle/appcenter/tasks/AppCenterUploadTask.groovy (hunk 1 above)
Comment (Larusso) [reply src/main/groovy/wooga/gradle/appcenter/tasks/AppCenterUploadTask.groovy:172]: The provider API just wraps the properties in a generic wrapper object which can hold either concrete values or other provider objects which all get evaluated at runtime when the value is requested. The properties itself can't no longer be simple java beans (getter/setter) but are `final` objects so they get constructed once for each instance. This code in the constructor is just the bootstrap. In grovvy it is possible to do that directly after the declaration like

```groovy
final Property<String> apiToken = project.objects.property(String)
```

But I personaly like the java way better. This also allows me to chain some provider together.
`project.object` just returns a factory with utility functions to create these generic `Property<T>`. One needs to pass the type of the property since java has not type inference.

File: src/main/groovy/wooga/gradle/compat/ProjectLayout.groovy (hunk 2 above)
Comment (Larusso) [reply src/main/groovy/wooga/gradle/compat/ProjectLayout.groovy:37]: One of my major issues with gradle is how they create and stabalize APIs.
The provider API was `@incubating` in gradle 4 and 5. I didn't check in gradle 6 but what they did is to move two property factory methods. Since gradle 6 is not API compatible we have to somehow make sure to keep the compatibility. Because we use gradle 4.10.2 for our projects and now also gradle 6.4 for android we need to stay compatible across all versions.

The methods `directoryProperty` and `fileProperty` used to be located in `ProjectLayout` instance. I honestly never liked this because you had to remember that these two properties had to be constructed from a different factory. So the gradle developers thought so too and moved these two methods into the `ObjectFactory` instance with gradle 6. I created this little helper class to call at runtime the correct location for the methods. When we move to gradle 6 for good we can remove this construct. I might move this out into a helper lib if I see the need in another plugin later.

**Comment by Larusso** [comment]
@alexandershestakov THX
